crates/ruff_cli/src/cache.rs

```diff
@@ -296,3 +296,164 @@ pub(crate) fn init(path: &Path) -> Result<()> {
+            let entry = entry.unwrap();
+            if !entry.file_type().unwrap().is_dir() {
+                continue;
+            }
+
+            for entry in fs::read_dir(entry.path()).unwrap() {
+                let entry = entry.unwrap();
+                if !entry.file_type().unwrap().is_file() {
+                    continue;
+                }
+
+                let path = entry.path();
+                expected_diagnostics += lint_path(
+                    &path,
+                    Some(&package_root),
+                    &settings,
+                    Some(&cache),
+                    flags::Noqa::Enabled,
+                    flags::FixMode::Generate,
+                )
+                .unwrap();
+                paths.push(path);
+            }
+        }
+        assert!(paths.len() >= 1, "no files checked");
+
+        cache.store().unwrap();
+
+        let cache = Cache::open(&cache_dir, package_root.to_owned(), &settings.lib);
+        assert!(cache.package.files.len() >= 1);
+
+        let mut got_diagnostics = Diagnostics::default();
+        for path in paths {
+            got_diagnostics += lint_path(
+                &path,
+                Some(&package_root),
+                &settings,
+                Some(&cache),
+                flags::Noqa::Enabled,
+                flags::FixMode::Generate,
+            )
+            .unwrap();
+        }
+
+        // Not stored in the cache.
+        expected_diagnostics.source_kind.clear();
+        got_diagnostics.source_kind.clear();
+        assert!(expected_diagnostics == got_diagnostics);
+    }
+
+    #[test]
+    fn invalidation() {
+        const SOURCE: &[u8] = b"a = 1\n\n__all__ = list([\"a\", \"b\"])\n";
+
+        let mut cache_dir = temp_dir();
+        cache_dir.push("ruff_tests/cache_invalidation");
+        let _ = fs::remove_dir_all(&cache_dir);
+        cache::init(&cache_dir).unwrap();
+
+        let settings = AllSettings::default();
+        let package_root = fs::canonicalize(&cache_dir).unwrap();
+        let cache = Cache::open(&cache_dir, package_root.to_owned(), &settings.lib);
+        assert_eq!(cache.new_files.lock().unwrap().len(), 0);
+
+        let path = cache_dir.join("source.py");
+        fs::write(&path, SOURCE).unwrap();
+
+        let mut expected_diagnostics = lint_path(
+            &path,
+            Some(&package_root),
+            &settings,
+            Some(&cache),
+            flags::Noqa::Enabled,
+            flags::FixMode::Generate,
+        )
+        .unwrap();
+        assert!(cache.new_files.lock().unwrap().len() == 1);
+
+        cache.store().unwrap();
+        // On certain tmp filesystems (tmpfs) overwriting the file with the same
+        // content doesn't seem to update it's modified timestamp. Calling fsync
+        // doesn't mark a difference, but this small sleep seems to fix the
+        // issue.
+        std::thread::sleep(std::time::Duration::from_millis(10));
```

Comment:
Can we set the last modification time manually instead of falling back to a sleep here?

Reply:
It's not about setting the last modified time. For some reason that tmpfs on my system didn't sync/write/whatever the timestamp correctly. So when the loop below started and the file was rewritten (the "File change" test case below) the timestamp was the same as in the cache (and thus the cache was reused causing an error). I didn't really bother investigating it more than that though.